snapshots/windows/cimfs.go

```diff
@@ -0,0 +1,232 @@
+type cimFSSnapshotter struct {
+	*windowsBaseSnapshotter
+	// cimDir is the path to the directory which holds all of the layer cim files.  CimFS needs all the
+	// layer cim files to be present in the same directory. Hence, cim files of all the snapshots (even if
+	// they are of different images) will be kept in the same directory.
+	cimDir string
+}
+
+func init() {
+	registry.Register(&plugin.Registration{
+		Type: plugins.SnapshotPlugin,
+		ID:   "cimfs",
+		InitFn: func(ic *plugin.InitContext) (interface{}, error) {
+			ic.Meta.Platforms = []ocispec.Platform{platforms.DefaultSpec()}
+			return NewCimFSSnapshotter(ic.Properties[plugins.PropertyRootDir])
+		},
+	})
+}
+
+// NewCimFSSnapshotter returns a new CimFS based windows snapshotter
+func NewCimFSSnapshotter(root string) (snapshots.Snapshotter, error) {
+	if !cimfs.IsCimFSSupported() {
+		return nil, fmt.Errorf("host windows version doesn't support CimFS")
+	}
+
+	baseSn, err := newBaseSnapshotter(root)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cimFSSnapshotter{
+		windowsBaseSnapshotter: baseSn,
+		cimDir:                 filepath.Join(baseSn.info.HomeDir, "cim-layers"),
+	}, nil
+}
+
+// getCimLayerPath returns the path of the cim file for the given snapshot. Note that this function doesn't
+// actually check if the cim layer exists it simply does string manipulation to generate the path isCimLayer
+// can be used to verify if it is actually a cim layer.
+func getCimLayerPath(cimDir, snID string) string {
+	return filepath.Join(cimDir, (snID + ".cim"))
+}
+
+// isCimLayer checks if the snapshot referred by the given key is actually a cim layer.  With CimFS
+// snapshotter all the read-only (i.e image) layers are stored in the cim format while the scratch layers (or
+// writable layers) are same as standard windows snapshotter scratch layers.
+func (s *cimFSSnapshotter) isCimLayer(ctx context.Context, key string) (bool, error) {
+	id, _, _, err := storage.GetInfo(ctx, key)
+	if err != nil {
+		return false, fmt.Errorf("get snapshot info: %w", err)
+	}
+	snCimPath := getCimLayerPath(s.cimDir, id)
+	if _, err := os.Stat(snCimPath); err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (s *cimFSSnapshotter) Usage(ctx context.Context, key string) (snapshots.Usage, error) {
+	baseUsage, err := s.windowsBaseSnapshotter.Usage(ctx, key)
+	if err != nil {
+		return snapshots.Usage{}, err
+	}
+
+	ctx, t, err := s.ms.TransactionContext(ctx, false)
+	if err != nil {
+		return snapshots.Usage{}, err
+	}
+	defer t.Rollback()
+
+	id, _, _, err := storage.GetInfo(ctx, key)
+	if err != nil {
+		return snapshots.Usage{}, fmt.Errorf("failed to get snapshot info: %w", err)
+	}
+
+	if ok, err := s.isCimLayer(ctx, key); err != nil {
+		return snapshots.Usage{}, err
+	} else if ok {
+		cimUsage, err := cimfs.GetCimUsage(ctx, getCimLayerPath(s.cimDir, id))
```

Comment:
This will only return the total space used by the region/objectID files of that CIM. It doesn't include the space used by the parents in case of a forked CIM.